### PR TITLE
[Model] CLIP Embedding Support

### PR DESCRIPTION
**Pull request by DarkLight1337**
## Purpose

Support CLIP text and image embedding in the same model.
- For text inputs, we only apply `token_embedding` when calling `get_input_embeddings`. The rest of the text embedding and the encoder logic are applied when calling `forward` on the model.
- For image inputs, we apply vision embeddings when calling `get_input_embeddings`. Since the model doesn't have a decoder, we directly return the embeddings inside the `forward` method.
- In dummy run, `forward` method doesn't receive image inputs so we cannot use the existence of `pixel_values` to determine whether `input_embeds` is from text or image inputs. To work around this, I have added a state `self._is_text_input` that is set inside `get_input_embeddings`. @ywang96 should we update `GPUModelRunner._dummy_mm_kwargs` to create the dummy multi-modal inputs for all multi-modal models?
- Strictly speaking, for CLIP, the pooling type is `LAST` for text, and `CLS` for image. But to simplify the code, we treat `LAST` pooling type for image inputs as `CLS` pooling type so that we don't have to define separate pooling types for now.

After this PR, it should be relatively straightforward to extend to SigLIP.

cc @maxdebayser @noooop 

FIX (partial) https://github.com/vllm-project/vllm/issues/25581

## Todo list

- [x] Add examples for offline inference and online serving
- [ ] Benchmarks

## Test Plan

- Added tests specific to CLIP model, which pass locally.

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/model_executor/models/clip.py [hunk 1]
```diff
@@ -260,6 +510,128 @@ def forward(
...
+                weight_loader = param.weight_loader
+                weight_loader(param, loaded_weight, shard_id)
+                break
+            else:
+                param = params_dict[name]
+                weight_loader = getattr(param, "weight_loader",
+                                        default_weight_loader)
+                weight_loader(param, loaded_weight)
+            loaded_params.add(name)
+        return loaded_params
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/model_executor/models/clip.py:597]: `AutoWeightsLoader` calls `load_weights` of inner modules

File: vllm/model_executor/models/clip.py [hunk 2]
```diff
@@ -260,6 +510,128 @@ def forward(
...
+        self,
+        input_ids: Optional[torch.Tensor],
+        position_ids: torch.Tensor,
+        inputs_embeds: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+        return self.text_model(
+            input_ids=input_ids,
+            position_ids=position_ids,
+            inputs_embeds=inputs_embeds,
+        )
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/model_executor/models/clip.py]: Removed

File: vllm/model_executor/models/clip.py [hunk 3]
```diff
@@ -45,7 +78,211 @@ def get_patch_grid_length(self) -> int:
...
+    def image_token_id(self) -> int:
+        tokenizer = self.info.get_tokenizer()
+        dummy_token_id = 0
+
+        assert (tokenizer.convert_ids_to_tokens([dummy_token_id])
+                not in tokenizer.special_tokens_map.values())
+
+        return dummy_token_id
+
+    def apply(
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/model_executor/models/clip.py:185]: Done

File: docs/models/supported_models.md [hunk 4]
```diff
@@ -827,6 +827,7 @@ The following table lists those that are tested in vLLM.
 
 | Architecture | Models | Inputs | Example HF Models | [LoRA](../features/lora.md) | [PP](../serving/parallelism_scaling.md) | [V1](gh-issue:8779) |
 |--------------|--------|--------|-------------------|----------------------|---------------------------|---------------------|
+| `CLIPModel` | CLIP | T + I | `openai/clip-vit-base-patch32`, `openai/clip-vit-large-patch14`, etc. | |  | ✅︎ |
```
Comment (Isotr0py) [inline docs/models/supported_models.md]:
```suggestion
| `CLIPModel` | CLIP | T / I | `openai/clip-vit-base-patch32`, `openai/clip-vit-large-patch14`, etc. | |  | ✅︎ |
```

File: vllm/model_executor/models/clip.py [hunk 5]
```diff
@@ -223,10 +469,12 @@ class CLIPEncoder(nn.Module):
 
     def __init__(
         self,
-        config: CLIPVisionConfig,
+        config: Union[CLIPTextConfig, CLIPVisionConfig],
         quant_config: Optional[QuantizationConfig] = None,
         num_hidden_layers_override: Optional[int] = None,
+        *,
         prefix: str = "",
+        attn_cls: Union[type[Attention], type[MultiHeadAttention]],
```
Comment (Isotr0py) [inline vllm/model_executor/models/clip.py:482]: IIRC, Clip's text encoder's attention is prefill-only casual attention. Does decoder's kv-cache really work with this text tower?

Perhaps we should decouple bidirectional mask with encoder-only attention to allow encoder-only causal attention. 🤔

File: docs/models/supported_models.md (hunk 4 above)
Comment (DarkLight1337) [reply docs/models/supported_models.md]: Nice catch

File: vllm/model_executor/models/clip.py (hunk 5 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/clip.py:482]: Indeed it should be `ENCODER_ONLY` causal attention, let me fix that.

Comment (DarkLight1337) [reply vllm/model_executor/models/clip.py:482]: Actually, I see several other pooling models use `DECODER_ONLY` attention in the language backbone when `is_causal=True`. I think using `DECODER_ONLY` attention to indicate causal mask should still work because pooling models don't have decode phase anyway.

Comment (DarkLight1337) [reply vllm/model_executor/models/clip.py:482]: Maybe @maxdebayser @heheda12345 have a better idea about this?

File: vllm/model_executor/models/bert.py [hunk 6]
```diff
@@ -351,7 +351,7 @@ def __init__(
                                    prefix=f"{prefix}.encoder")
 
     def get_input_embeddings(self, input_ids: torch.Tensor) -> torch.Tensor:
-        return self.embeddings(input_ids)
+        return self.embeddings.word_embeddings(input_ids)
```
Comment (maxdebayser) [inline vllm/model_executor/models/bert.py:354]: Why is this change required?

File: vllm/model_executor/models/clip.py (hunk 5 above)
Comment (maxdebayser) [reply vllm/model_executor/models/clip.py:482]: But if the attention is causal, it wouldn't hurt to use `DECODER_ONLY`, right? In addition it can enable chunked prefill and prefix caching if the pooling type supports it.

Comment (DarkLight1337) [reply vllm/model_executor/models/clip.py:482]: Yes, currently the PR is using `DECODER_ONLY` which is the default attention type

File: vllm/model_executor/models/bert.py (hunk 6 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/bert.py:354]: > For text inputs, we only apply token_embedding when calling get_input_embeddings. The rest of the text embedding and the encoder logic are applied when calling forward on the model.

For prompt embedding inputs to work, I figured out this flow in this PR, so I applied it to BERT as well

File: vllm/model_executor/models/clip.py (hunk 5 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/clip.py:482]: Would that cause problems since the text backbone is actually an encoder?

Comment (maxdebayser) [reply vllm/model_executor/models/clip.py:482]: I could be mistaken, but for text the core difference between encoder and decoder is the attention mask. The other differences are mainly a consequence of this. For example, with the bi-directional cache attention it doesn't make a lot of sense to use KV cache, hence in vllm we've refactored the Bert models and other to use the `EncoderOnlyAttention` class. So if the CLIP attention is causal, it should work fine with the decoder attention.

Comment (DarkLight1337) [reply vllm/model_executor/models/clip.py:482]: Ok cool, let's merge this then

**Comment by DarkLight1337** [comment]
@Isotr0py can you approve this?

**Review by Isotr0py — APPROVED** [review]
